### PR TITLE
CB-11302: Allow the ability for customer to choose a previously relea…

### PR DESCRIPTION
**Pull request by LinaAtAustin**
…sed hotfix

-The approach is to add a new input in cdp cli. So customers can specify the image catalog and uuid when creating the data lake
-add a new field in the SdxClusterRequest that contains the image info
-the image info is set to StackV4Request

See detailed description in the commit message.

### CODE REVIEW

**Comment by LinaAtAustin** [comment]
I have run the local deployment to test by hardcoding the imagine info the SdxClusterRequest. However, I kept on getting external database deployment issue. Not sure if it is caused by the cost saving script or the image info I used {"catalog":"cloudbreak-default","id":"a0b2b750-9845-42e0-526f-fcba813f33ac"}. Will test again later

**Comment by LinaAtAustin** [comment]
Screen Shot 2021-02-24 at 12 05 49 AM

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequest.java [hunk 1]
```diff
@@ -29,6 +30,8 @@
 
     private boolean enableRangerRaz;
 
+    private ImageSettingsV4Request imageSettingsV4Request;
```
Comment (sodre90) [inline datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequest.java:33]: Please create a separate endpoint with a request like "CustomSdxClusterRequest". The runtime version is meaningless for this request, so we can create an ancestor class for SdxClusterRequest and CustomSdxClusterRequest without runtime version field. There is some logic for runtime version and it should be reviewed.

Comment (akanto) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequest.java:33]: I am also surprised that sdx-api is able reference classes from core-api

**Comment by akanto** [comment]
This pull request requires junit tests.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java [hunk 2]
```diff
@@ -309,6 +309,7 @@ private StackV4Request getStackRequest(SdxClusterRequest sdxClusterRequest, Stac
                         ", runtime version: " + runtimeVersion);
             }
             stackRequest.getCluster().setRangerRazEnabled(sdxClusterRequest.isEnableRangerRaz());
+            stackRequest.setImage(sdxClusterRequest.getImageSettingsV4Request());
```
Comment (sodre90) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:312]: What if somebody uses internal api and sends a stackv4request without image info and would like to use image provided in image field? It should be handled in 'else' branch.

Comment (LinaAtAustin) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:312]: OK. I will handle that scenario as well. Thanks!

**Comment by sodre90** [comment]
> This pull request requires junit tests.

I think e2e mock tests also required. 

This new field has risks. Until now cloudbreak core decided which image will be used to create the datalake stack but it won't be true anymore. We tried not to expose image id to datalake API because it can cause many support cases as customers can specify such dumb things like aws image for azure environment or base image etc. It was allowed only for internal use for QE.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequest.java (hunk 1 above)
Comment (LinaAtAustin) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequest.java]: introducing a base class, which does not have "runtime" field requires changing a lot of places. So I derive the SdxCustomClusterRequest from SdxClusterRequest and validate that the runtime field and image field cannot be set at the same time.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java (hunk 2 above)
Comment (LinaAtAustin) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java]: Handle that for both cases

**Comment by LinaAtAustin** [comment]
I will create a new entitlement in CB-11344. Once that is committed, I will add entitlement check in SdxService to allow custom request only when the entitlement is on.

I have added several unit tests. Will run end-to-end test tomorrow.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.request.StackV4Request;
 
-public class SdxInternalClusterRequest extends SdxClusterRequest {
+public class SdxInternalClusterRequest extends SdxCustomClusterRequest {
```
Comment (lajosrodek) [inline datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java:5]: I wonder if it has any benefit (or dare I say, if it makes sense) to allow an extra image UUID override for the internal request. In particular, `SdxInternalClusterRequest` already contains a complete `StackV4Request` that may or may not have an image UUID inside. In any case, I do not think it is a good idea to override the ID inside the `StackV4Request` with the one coming from SdxCustomClusterRequest.

File: datalake/src/main/java/com/sequenceiq/datalake/controller/sdx/SdxCustomController.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    @Inject
+    private SdxMetricService metricService;
+
+    @Override
+    @CheckPermissionByAccount(action = AuthorizationResourceAction.CREATE_DATALAKE)
+    public SdxClusterResponse create(String name, @Valid SdxCustomClusterRequest createSdxClusterRequest) {
+        String userCrn = ThreadBasedUserCrnProvider.getUserCrn();
+        Pair<SdxCluster, FlowIdentifier> result = sdxService.createSdx(userCrn, name, createSdxClusterRequest, null);
+        SdxCluster sdxCluster = result.getLeft();
+        metricService.incrementMetricCounter(MetricType.INTERNAL_SDX_REQUESTED, sdxCluster);
```
Comment (lajosrodek) [inline datalake/src/main/java/com/sequenceiq/datalake/controller/sdx/SdxCustomController.java:39]: Do not we possibly need some either kind of metric here?

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java [hunk 5]
```diff
@@ -190,6 +202,32 @@ public SdxCluster getByCrn(String userCrn, String clusterCrn) {
...
     }
 
+    public String getRuntimeVersionFromImageRequest(ImageSettingsV4Request imageSettingsV4Request) {
+        try {
+            LOGGER.info("Calling cloudbreak to get runtime version for the given image catalog {} and image id {}",
+                    imageSettingsV4Request.getCatalog(), imageSettingsV4Request.getId());
+            ImageV4Response imageV4Response =  ThreadBasedUserCrnProvider.doAsInternalActor(
+                    () -> {
+                        try {
+                            return imageCatalogV4Endpoint.getImageByNameAndId(0L, imageSettingsV4Request.getCatalog(), imageSettingsV4Request.getId());
```
Comment (lajosrodek) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:212]: Not a big deal but could you please extract `0L` as a new constant `WORKSPACE_ID`?

**Comment by LinaAtAustin** [comment]
@sodre90 "This new field has risks. " I have what cloud platform is and I get ImageV4Response using UUID to know the runtime. So I can check if the platform matches (ImageV4Response.imageSetsByProvider contains the mapping from cloud provider to the image sets). If you want to check other items, I can add those validation before creating the stack. In this way, we can achieve both flexibility and avoid wrong configuration as much as possible. 

If customer wants to use an image with Its UUID, I know it is not available when getting its info. So the DL creation can fail fast. customer can use a different image or request to put that image in CDP. Ideally, each customer can have its own image catalog, and Cloudera can charge for maintaining such additional service. The storage cost should be much smaller than what customers want to pay to get this service.

I will add mock tests.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java (hunk 3 above)
Comment (LinaAtAustin) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java:5]: if ((stackRequest.getImage() == null) && (sdxClusterRequest instanceof SdxCustomClusterRequest)) {
            stackRequest.setImage(((SdxCustomClusterRequest) sdxClusterRequest).getImageSettingsV4Request());
}

You can see the overwrite happens only if the StackV4Request contains no image settings. Are you still concerned? Peter wants to set image when user specifies it in internal request. I am trying to satisfy both opinions

File: datalake/src/main/java/com/sequenceiq/datalake/controller/sdx/SdxCustomController.java (hunk 4 above)
Comment (LinaAtAustin) [reply datalake/src/main/java/com/sequenceiq/datalake/controller/sdx/SdxCustomController.java:39]: I added a new metric type CUSTOM_SDX_REQUESTED to record the metric

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java (hunk 5 above)
Comment (LinaAtAustin) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:212]: done

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java (hunk 3 above)
Comment (lajosrodek) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java:5]: Yes, I have seen the above logic but I think it shall not touch the image ID in `SdxInternalClusterRequest` whatsoever. If the user sending the req omitted the image ID, he/she probably wants to leave it that way. If he/she wanted to have an image ID explicitly, that could have been included in `SdxInternalClusterRequest.stackV4Request`, without relying on the UUID override of `SdxCustomClusterRequest`.

In short, my feeling is `SdxInternalClusterRequest` shall keep `SdxClusterRequest` as its superclass and not have the extras of `SdxCustomClusterRequest`. The logic you quoted will still work in both cases, but it will override the `null image` only when using `SdxCustomClusterRequest` instead of `SdxInternalClusterRequest`.

@sodre90 @doktoric @lnardai Please do not be shy and share your thoughts. 😃

Comment (lajosrodek) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java:5]: Actually, thinking a bit more I would even go further: I think `SdxCustomClusterRequest` should be a type not related to either `SdxClusterRequest` or `SdxInternalClusterRequest`. The cleanest approach would be to make `SdxCustomClusterRequest` be a subclass of `Object`, having all the fields of `SdxClusterRequest` except for the `runtime`. That way the conflict between `runtime` and `SdxCustomClusterRequest.imageSettingsV4Request` could be avoided entirely.

Comment (LinaAtAustin) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java:5]: @lajosrodek I prefer to not overwrite `image` in `SdxCustomClusterRequest.` In this way, user has full control of its value. Adding additional logic internally could be confusing to user. @sodre90 Are you OK with this? User who issues internal CLI can setting the image, why not let user get what they set directly?

@lajosrodek In the internal CLI, we have the same issue of conflict between `SdxInternalClusterRequest.StackV4Request.runtime` and `SdxInternalClusterRequest.ImageSettingsV4Request`. 

`The cleanest approach would be to make SdxCustomClusterRequest be a subclass of Object, having all the fields of SdxClusterRequest except for the runtime`
How about the following approach?
1. I define SdxCustomClusterRequest as you mentioned
2. then In SdxCustomController, I create `SdxClusterRequest` whose `runtime` is null and add a new input `ImageSettingsV4Request` for SdxService.createSdx(). In this way, we can avoid the conflict totally, and still minimize the code changes in many places.

Comment (lajosrodek) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java:5]: @LinaAtAustin Yes, `SdxInternalClusterRequest` should have also used `Object` as its superclass instead of `SdxClusterRequest`. The `SdxClusterRequest.runtime` is in conflict with `SdxInternalClusterRequest.StackV4Request` because the latter may prescribe a totally different runtime & image combination. We cannot help that here easily.

Instead, what I suggested are
1. Not to make `SdxInternalClusterRequest` be a subclass of `SdxCustomClusterRequest`; and
2. Make `SdxCustomClusterRequest` be a subclass of `Object`.

The controller change (i.e. converting `SdxCustomClusterRequest` into `SdxClusterRequest` and passing the additional `ImageSettingsV4Request` to `SdxService.createSdx()` sounds good. 👍  

Let us wait for some more comments as the rest of the team has not yet expressed its opinions.

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/v4/ImageCatalogV4Controller.java [hunk 6]
```diff
@@ -142,4 +144,11 @@ public ImagesV4Response getImagesByName(Long workspaceId, @ResourceName String n
         Images images = imageCatalogService.getImagesByCatalogName(workspaceId, name, stackName, platform);
         return converterUtil.convert(images, ImagesV4Response.class);
     }
+
+    @Override
+    @CheckPermissionByResourceName(action = AuthorizationResourceAction.DESCRIBE_IMAGE_CATALOG)
+    public ImageV4Response getImageByNameAndId(Long workspaceId, @ResourceName String name, String imageId) throws Exception {
```
Comment (sodre90) [inline core/src/main/java/com/sequenceiq/cloudbreak/controller/v4/ImageCatalogV4Controller.java:150]: can we rename it to getImageByCatalogNameAndImageId to make its name more specific?

Comment (LinaAtAustin) [reply core/src/main/java/com/sequenceiq/cloudbreak/controller/v4/ImageCatalogV4Controller.java:150]: sure. It is done

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java (hunk 3 above)
Comment (sodre90) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java:5]: @LinaAtAustin We would like to avoid any "instanceof". As I see we have too many parameters that can be null, so I understand your point and your code, but as of now, I would recommend if we just add the ImageSettingsV4Request as a new parameter to createSdx method. It is ugly, but it is the easiest way for now and I will create a ticket for us to refactor / rearrange createSdx method.

I think we should create an abstract ancestor class for SdxClusterRequest like SdxClusterRequestBase and put all its parameters from SdxClusterRequest except runtime. SdxClusterRequest and SdxCustomClusterRequest should extend SdxClusterRequestBase. So now we have to add one more parameter to createSdx, the runtime field, and we have to change sdxClusterRequest paramater's class to SdxClusterRequestBase. I admit it is way more ugly now, we added 2 more parameters to createSdx method but on the API side we are fine. Maybe we can create a parameter class with parameters of createSdx but it is not a big improvement :) (userCrn, name, sdxClusterRequestBase, runtime, internalStackV4Request, imageSettingsV4Request).

You can implement your code without instanceOf, but with some null checks (our instanceOf hater soul will be in peace) and we will be fine on API side (no runtime field in custom request). The code is very ugly at the moment (the original one), it will be a little uglier but we can explode this method later, I will create a jira about this because I don't want to push this work to you. We were the ones who created this mess, it is our task to clean it.

I wouldn't touch SdxInternalClusterRequest (I think it shouldn't have runtime anyway, but if we remove it then we will break the API compatibility..)

Comment (lajosrodek) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java:5]: Good point, using a common ancestor for `SdxClusterRequest` and `SdxCustomClusterRequest` to extract common fields makes more sense. 😄

Comment (LinaAtAustin) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java:5]: I agree. Will update soon

**Comment by LinaAtAustin** [comment]
@lajosrodek @sodre90 Can you take a look at the update? I have changed the API based on our discussion. I have not got time to add mock end-to-end test since I have been working on customer escalation since Friday and over the weekend. Thanks!

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequestBase.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+    @NotNull
+    private SdxClusterShape clusterShape;
+
+    private SdxCloudStorageRequest cloudStorage;
+
+    private SdxDatabaseRequest externalDatabase;
+
+    @Valid
+    private SdxAwsRequest aws;
```
Comment (lajosrodek) [inline datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequestBase.java:24]: I know this is not part of the PR, but perhaps it made sense to add a proper `toString()` implementation for the classes involved here. `SdxClusterShape` is an enum, so it will work as expected, but the other three types will show the default (and pretty useless) `Object.toString()` output.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    public String getRuntime() {
+        return null;
+    }
+
+    @Override
+    public String toString() {
+        String baseString = super.toString();
+        String additionalString = ", imageSettingsV4Request=" + imageSettingsV4Request +  '}';
+        String ownString = baseString.replace("SdxClusterRequestBase", "SdxCustomClusterRequest");
+        return ownString.replace("}", additionalString);
```
Comment (lajosrodek) [inline datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java:29]: Like in `SdxClusterRequest`, I do not think it is a good idea to resort to string-wrestling.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    public Pair<SdxClusterRequest, ImageSettingsV4Request> convert() throws IOException {
```
Comment (lajosrodek) [inline datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java:32]: Please use a more specific name, such as `asPair()`, `toPair()` or `convertToPair()`.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            }
+
+            @Override
+            public ImageSettingsV4Request getRight() {
+                return getImageSettingsV4Request();
+            }
+
+            @Override
+            public ImageSettingsV4Request setValue(ImageSettingsV4Request value) {
+                return value;
```
Comment (lajosrodek) [inline datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java:55]: Or perhaps throw `UnsupportedOperationException`?

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java [hunk 11]
```diff
@@ -230,13 +277,30 @@ public SdxCluster getByNameInAccount(String userCrn, String name) {
...
+            throw new BadRequestException("SDX customer request is not valid.", ex);
+        }
+    }
+
     public Pair<SdxCluster, FlowIdentifier> createSdx(final String userCrn, final String name, final SdxClusterRequest sdxClusterRequest,
             final StackV4Request internalStackV4Request) {
+        return createSdx(userCrn, name, sdxClusterRequest, internalStackV4Request, null);
+    }
+
+    protected Pair<SdxCluster, FlowIdentifier> createSdx(final String userCrn, final String name, final SdxClusterRequest sdxClusterRequest,
```
Comment (lajosrodek) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:295]: I feel this should be `private`.

**Review by lajosrodek — CHANGES_REQUESTED** [review]
Thanks, looks alright. Some further minor comments.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequest.java [hunk 12]
```diff
@@ -53,71 +12,10 @@ public void setRuntime(String runtime) {
...
-                "clusterShape=" + clusterShape +
-                ", runtime='" + runtime + '\'' +
-                ", cloudStorage=" + cloudStorage +
-                ", externalDatabase=" + externalDatabase +
-                ", aws=" + aws +
-                ", enableRangerRaz=" + enableRangerRaz +
-                '}';
+        String baseString = super.toString();
+        String ownString = baseString.replace("SdxClusterRequestBase", "SdxClusterRequest");
+        return ownString.replace("}", ", runtime='" + runtime + '}');
```
[the remark this answers is not in the document]
Comment (LinaAtAustin) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequest.java:19]: that is good point. I will update. Thanks!

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java (hunk 8 above)
Comment (LinaAtAustin) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java:29]: I will update accordingly.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java (hunk 9 above)
Comment (LinaAtAustin) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java:32]: I changed to convertToPair()

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java (hunk 10 above)
Comment (LinaAtAustin) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxCustomClusterRequest.java:55]: I changed it to "throw exception "

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java (hunk 11 above)
Comment (LinaAtAustin) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:295]: done

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequestBase.java (hunk 7 above)
Comment (LinaAtAustin) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequestBase.java:24]: I have created CB-11443 to track that effort. So I can focus on the feature in this PR.

**Review by lajosrodek — APPROVED** [review]
LGTM nice work!

**Comment by LinaAtAustin** [comment]
@sodre90 I have added mock test. Can you take a look? Thanks!

**Comment by LinaAtAustin** [comment]
DataLake created with custom image
The image catalog and ID
I enabled the entitlement CDP_DATA_LAKE_CUSTOM_IMAGE, set the catalog to be "cloudbreak-default" and image id to be "b512d50c-44ee-4e87-a5a3-a308ead95086". DL was created successfully using the custom image.

**Review by topolyai5 — COMMENTED** [review]
Is this feature an internal or external feature?

if internal why don't you use the internal endpoint? if external why don't you integrate to the existing endpoint? Is it really need?

**Comment by LinaAtAustin** [comment]
@topolyai5 
It is for beta cli. So it is not internal. It is hidden from normal customer. The reason for this to be in beta CLI is because we don't want to expose this to all customers, and would like only for PMs to use it for select customers.
The way to do this is to set the x-hidden-retention property in the yaml.. CLI update is tracked in https://jira.cloudera.com/browse/CDPCP-4045

@perdos suggested to use another endpoint as the request contains image settings, to avoid conflicting info, it does not have "runtime" field. This is to prevent mistake that customer could make and cause issue.

**Review by topolyai5 — CHANGES_REQUESTED** [review]
As we talked in Slack channel, could you please move the endpoint operation to the common controller as an operation: /sdx/custom
I suggest, rename it to a more meaningful name, that refers to the operation

**Comment by LinaAtAustin** [comment]
@topolyai5 @sodre90  
Peter, you have mentioned in https://github.com/hortonworks/cloudbreak/pull/10124#discussion_r581762031 and asked me to "Please create a separate endpoint with a request like "CustomSdxClusterRequest". The runtime version is meaningless for this request, so we can create an ancestor class for SdxClusterRequest and CustomSdxClusterRequest without runtime version field. There is some logic for runtime version and it should be reviewed."

Gergely is asking me to not create a separate endpoint, and move it back to the common controller. These are conflicting requests.
Can you discuss and decide what way I should go to finish this code change? I have very limited time before I am away next week.

@topolyai5 how important is it to remove this new endpoint? The concern of moving it to the same common controller is that new we have two different types of requests that contain different information to the same endpoint. It may be easier for customer to make mistake. 

The new feature in this code change is not intend to be open to the most customers as Karthik mentioned in https://jira.cloudera.com/browse/CDPCP-4045

having separate endpoint makes it easy to hide this feature, and only allow limited customer to access it.

**Comment by LinaAtAustin** [comment]
Gergely, I saw you and Peter reached agreement. I am making this change now.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/endpoint/SdxEndpoint.java [hunk 13]
```diff
@@ -255,4 +256,10 @@ RangerCloudIdentitySyncStatus getRangerCloudIdentitySyncStatus(@PathParam("envCr
     @Produces(MediaType.APPLICATION_JSON)
     @ApiOperation(value = ROTATE_CERTIFICATES, nickname = "rotateAutoTlsCertificatesByCrn")
     FlowIdentifier rotateAutoTlsCertificatesByCrn(@PathParam("crn") String crn, @Valid CertificatesRotationV4Request rotateCertificateRequest);
+
+    @POST
+    @Path("/customimage/{name}")
```
Comment (topolyai5) [inline datalake-api/src/main/java/com/sequenceiq/sdx/api/endpoint/SdxEndpoint.java:261]: Can you please follow the pattern? we use this pattern: `{name}/operation_name` or `{crn}/operation_name`. Our URL parser won't recognise the pattern.

Comment (LinaAtAustin) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/endpoint/SdxEndpoint.java:261]: @topolyai5 Thanks for the comment. Will @Path("{name}/customimage") work? If so, I will make update tomorrow. 

Is this the only change I need to make? If not, can we have 30 minutes zoom session to discuss and get all changes we need at once, then I can work on it on Monday. Thanks!

Comment (topolyai5) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/endpoint/SdxEndpoint.java:261]: I mean for this path: `{name}/custom_image`

Comment (topolyai5) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/endpoint/SdxEndpoint.java:261]: maybe it can work with crn as well?

Comment (LinaAtAustin) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/endpoint/SdxEndpoint.java]: @topolyai5 I don't think it work with crn. This is create function. the caller can only provide the DL name. The crn is not assigned yet. To get the details, we can use the crn, and actually reuse the existing function to do so. 

The added new function is only create() with customized image info.

Comment (topolyai5) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/endpoint/SdxEndpoint.java]: yes, you are right

**Review by topolyai5 — APPROVED** [review]
